drivers/spi/spi_xec_qmspi_v2.c

```diff
@@ -0,0 +1,751 @@
+
+	qmspi_set_signalling_mode(regs, smode);
+
+	/* chip select */
+	smode = regs->MODE & ~(MCHP_QMSPI_M_CS_MASK);
+#if DT_INST_PROP(0, chip_select) == 0
+	smode |= MCHP_QMSPI_M_CS0;
+#else
+	smode |= MCHP_QMSPI_M_CS1;
+#endif
+	regs->MODE = smode;
+
+	/* chip select timing */
+	regs->CSTM = cfg->cs_timing;
+
+	data->ctx.config = config;
+
+	/* Add driver specific data to SPI context structure */
+	spi_context_cs_configure(&data->ctx);
+
+	regs->MODE |= MCHP_QMSPI_M_ACTIVATE;
+
+	return 0;
+}
+
+/*
+ * Transmit dummy clocks - QMSPI will generate requested number of
+ * SPI clocks with I/O pins tri-stated.
+ * Single mode: 1 bit per clock -> IFM field = 00b. Max 0x7fff clocks
+ * Dual mode: 2 bits per clock  -> IFM field = 01b. Max 0x3fff clocks
+ * Quad mode: 4 bits per clock  -> IFM fiels = 1xb. Max 0x1fff clocks
+ * QMSPI unit size set to bits.
+ */
+static int qmspi_tx_dummy_clocks(struct qmspi_regs *regs, uint32_t nclocks)
+{
+	uint32_t descr, ifm, qstatus;
+
+	ifm = regs->CTRL & MCHP_QMSPI_C_IFM_MASK;
+	descr = ifm | MCHP_QMSPI_C_TX_DIS | MCHP_QMSPI_C_XFR_UNITS_BITS
+		| MCHP_QMSPI_C_DESCR_LAST | MCHP_QMSPI_C_DESCR0;
+
+	if (ifm & 0x01) {
+		nclocks <<= 1;
+	} else if (ifm & 0x02) {
+		nclocks <<= 2;
+	}
+	descr |= (nclocks << MCHP_QMSPI_C_XFR_NUNITS_POS);
+
+	regs->DESCR[0] = descr;
+
+	regs->CTRL |= MCHP_QMSPI_C_DESCR_EN;
+	regs->IEN = 0;
+	regs->STS = 0xfffffffful;
+
+	regs->EXE = MCHP_QMSPI_EXE_START;
+	/* TODO timeout */
+	do {
+		qstatus = regs->STS;
+		if (qstatus & MCHP_QMSPI_STS_PROG_ERR) {
+			return -EIO;
+		}
+	} while ((qstatus & MCHP_QMSPI_STS_DONE) == 0);
+
+	return 0;
+}
+
+/*
+ * Return unit size power of 2 given number of bytes to transfer.
+ */
+static uint32_t qlen_shift(uint32_t len)
+{
+	uint32_t ushift;
+
+	/* is len a multiple of 4 or 16? */
+	if ((len & 0x0F) == 0) {
+		ushift = 4;
+	} else if ((len & 0x03) == 0) {
+		ushift = 2;
+	} else {
+		ushift = 0;
+	}
+
+	return ushift;
+}
+
+/*
+ * Return QMSPI unit size of the number of units field in QMSPI
+ * control/descriptor register.
+ * Input: power of 2 unit size 4, 2, or 0(default) corresponding
+ * to 16, 4, or 1 byte units.
+ */
+static uint32_t get_qunits(uint32_t qshift)
+{
+	if (qshift == 4) {
+		return MCHP_QMSPI_C_XFR_UNITS_16;
+	} else if (qshift == 2) {
+		return MCHP_QMSPI_C_XFR_UNITS_4;
+	} else {
+		return MCHP_QMSPI_C_XFR_UNITS_1;
+	}
+}
+
+/*
+ * Allocate(build) one or more descriptors.
+ * QMSPI contains 16 32-bit descriptor registers used as a linked
+ * list of operations. Using only 32-bits there are limitations.
+ * Each descriptor is limited to 0x7FFF units where unit size can
+ * be 1, 4, or 16 bytes. A descriptor can perform transmit or receive
+ * but not both simultaneously. Order of descriptor processing is specified
+ * by the first descriptor field of the control register, the next descriptor
+ * fields in each descriptor, and the descriptors last flag.
+ */
+static int qmspi_descr_alloc(struct qmspi_regs *regs, const struct spi_buf *txb,
```

Comment:
@scottwcpg  could you document the parameters of this internal function?  I'm assuming didx is the descriptor index 
if that's the case, maybe rework the parameter list and pass didx as reference.
Returning an error and also valid value may lead to error for future maintainers. 
Seems didx = 0 is not equivalent to success in all scenarios.